src/funnels/Swapper.sol

```diff
@@ -0,0 +1,243 @@
+        require(keepers[msg.sender] == 1, "Swapper/non-keeper"); 
+        _;
+    }
+
+    uint256 internal constant WAD = 10 ** 18;
+
+    address public immutable uniV3Router;
+    address public immutable nst;
+    address public immutable gem;
+
+    function rely(address usr)   external auth { wards[usr]   = 1; emit Rely(usr); }
+    function deny(address usr)   external auth { wards[usr]   = 0; emit Deny(usr); }
+    function kiss(address usr)   external auth { buds[usr]    = 1; emit Kissed(usr); }
+    function diss(address usr)   external auth { buds[usr]    = 0; emit Dissed(usr); }
+    function permit(address usr) external toll { keepers[usr] = 1; emit Permit(usr); }
+    function forbid(address usr) external toll { keepers[usr] = 0; emit Forbid(usr); }
+
+    function file(bytes32 what, uint256 data) external auth {
+        if      (what == "maxIn")   maxIn  = data;
+        else if (what == "maxOut")  maxOut = data;
+        else if (what == "wantIn")  wantIn = data;
+        else if (what == "wantOut") wantOut = data;
+        else if (what == "hop")     hop    = data;
+        else if (what == "fee")     fee    = data;
+        else revert("Swapper/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        if (what == "buffer") buffer = data;
+        else revert("Swapper/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 what, address data, uint256 val) external auth {
+        if (what == "box") boxes[data] = val;
+        else revert("Swapper/file-unrecognized-param");
+        emit File(what, data, val);
+    }
+
+    function setWeights(Weight[] memory newWeights) external toll {
+        uint256 cumPct;
+        uint256[] memory arr = new uint256[](newWeights.length);
+        for(uint256 i; i < newWeights.length;) {
+            (address _box, uint256 _pct) = (newWeights[i].box, uint256(newWeights[i].wad));
+            require(boxes[_box] == 1, "Swapper/invalid-box");
+            require(_pct <= WAD, "Swapper/not-wad-percentage");
+            arr[i] = (uint256(uint160(_box)) << 96) | _pct;
+            cumPct += _pct;
+            unchecked { ++i; }
+        }
+        require(cumPct == WAD, "Swapper/total-weight-not-wad");
+        weights = arr;
+    }
+
+    function getWeightAt(uint256 i) public view returns (address box, uint256 percent) {
+        uint256 weight = weights[i];
+        (box, percent) = (address(uint160(weight >> 96)), weight & (2**96 - 1)); // TODO: check that this tuple assignment results in only one SLOAD
+        require(boxes[box] == 1, "Swapper/invalid-box"); // sanity check that any deauthorised box has also been removed from the weights vector
+    }
+
+    function getWeightsLength() external view returns (uint256 len) {
+        len = weights.length;
+    }
+
+    function swapIn(uint256 wad, uint256 min) external keeper returns (uint256 out) {
```

Comment:
So curious to know what others think, but for me the "In" and "Out" notation is really confusing.
I mean `In` direction can be either the buffer or the conduits.

How about changing to `buyGem` and `sellGem` like the psm?

Reply:
The in/out was intended to be understood as the direction of the movements of funds relative to the boxes: "swapping and moving the funds _in_ or _out_ of the boxes". But I can try to rename things using the nst/gem reference if that's more intuitive.

Reply:
Yeah I think I slightly prefer something that is more explicit to what we are doing. `buyGem/sellGem` or `nstToGem/gemToNst`. One would match the PSM, the other the new converters.

Reply:
Changed to `nstToGem/gemToNst`